File: ruby/Driver.rb
require "benchmark"
require "Game"
require "ConsolePlayer"

class RandomOptionPlayer 
	def initialize(name)
		@name = name
	end
	def select_move(moves, board)
		return moves.sample
	end
	def to_s
		return @name
	end
end

def profile(numRuns)
    totalTime = 0
    
    player1 = RandomOptionPlayer.new("player 1")
    player2 = RandomOptionPlayer.new("player 2")
    winsPerPlayer = {player1 => 0, player2 => 0}
    (1..numRuns).each do |run|
		game = Game.new(player1, player2)
		totalTime += Benchmark.realtime { winsPerPlayer[game.play()] += 1 }
	end
    msPerGame = (totalTime*1000) / numRuns
    puts "Total time is %0.2f sec to play %d games, or %0.2f ms/game" % [totalTime, numRuns, msPerGame]
    puts "Wins per player: "
    winsPerPlayer.each do |p,v| 
		puts p.to_s + ": " + v.to_s
    end
end

def playInteractive
	Game.new(ConsolePlayer.new("Player 1"), ConsolePlayer.new("Player 2"), eventHandler: ConsoleEventHandler.new).play
end

if ARGV.length > 0 then
	profile(ARGV[0].to_i)
else
	playInteractive # TODO
end

File: perl/driver.pl
#!/usr/bin/perl

use Checkers::Game;
use Checkers::Board;
use Checkers::MoveRules;
use strict;
use warnings;

sub profile {
    my $numRuns = shift;
    
    use Time::HiRes;

    my $totalTime = 0;
    my $randomPlayer = { selectMove => sub { my ($b,@m) = @_; return $m[ rand @m ]; } }; bless $randomPlayer;
    
    for (my $i = 0; $i < $numRuns; $i += 1 ) {
        my $board = Checkers::Board->new;
        my $start = Time::HiRes::time;
        Checkers::Game::play( $board, $randomPlayer, $randomPlayer, \&Checkers::MoveRules::getMovesForColor, undef);
        my $end = Time::HiRes::time;
        $totalTime += ($end-$start);
    }
    
    my $msPerGame = ($totalTime*1000) / $numRuns;
    printf "Total time is %.2f sec to play %d games, or %.2f ms/game\n", $totalTime, $numRuns, $msPerGame;
}

sub playInteractive {
    use Checkers::Console;
    
    my $board = Checkers::Board->new;
    Checkers::Game::play( $board, Checkers::Console::player, Checkers::Console::player, \&Checkers::MoveRules::getMovesForColor, Checkers::Console::eventHandler);
}

if (scalar @ARGV > 0) {
    profile($ARGV[0]);
}
else {
    playInteractive();
}
